### ISSUE TITLE
Documentation issues

### ISSUE THREAD

**Issue by boomcan90** [ArmyCyberInstitute/cmgr#19]
Working with `cmgr`, there's a few issues that I've noted that I think should be documented: 

- Docker multi stage builds cannot have the name `builder`. `FROM ubuntu:20.04 as builder` for example complains about the `/challenge` directory not being present, which should not be the case as this would only be an intermediary building docker image.
- When publishing the port, everything except the name (`# PUBLISH <port> AS <name> `) needs to be capitalized. Since I couldn't see the reference to `PUBLISH` in dockerfile documentation [here](https://docs.docker.com/engine/reference/builder/), I presume that this is something from cmgr. 
- The `CMGR_INTERFACE` doesn't seem to be working. Using the setup below, as seen in the image, the last word is the templated var `{{server}}`, and its showing as localhost rather than the expected IP set in the env variables. 
```
$ export CMGR_INTERFACE=192.168.1.119
$ echo $CMGR_INTERFACE
192.168.1.119
$ sudo ./cmgr playtest picoctf/iot/certpinning
challenge information available at: http://localhost:4242/
```
![Screenshot from 2021-01-19 09-47-47](https://user-images.githubusercontent.com/10251359/105073089-722c2c80-5a3b-11eb-82ce-a5cb98034c62.png)

**Comment by jrolli** [comment]
Thank you for the issue.  When working on your challenge, did you reference the documentation in the examples directory at all (particularly the `README.md` in the for the "custom" example)?

- The `builder` issue is an undocumented overload by `cmgr` which is designed to allow you to identify a stage separate from the user-facing one for returning the `metadata.json` and `artifacts.tar.gz` (was needed to get toward a functioning multi-container solution).  My default would be to document this in the `README.md` for the custom example, but where would you have expected to see this documented?
- The `# PUBLISH` directives are a cmgr thing and are currently documented in the README mentioned above.  Would adding a comment about case sensitivity there have helped you or would expect that to have been documented else where?
- For the interface issue, can you confirm whether you are using the latest version of `cmgr`?  This is related to issue #18, and should be fixed as of `v0.8.7` which was released on 2021-07-01 (Note: `v0.8.8` is the current release).

**Comment by boomcan90** [comment]
Yes, I've looked through the examples directory (looking at the custom and multi examples). I think having them as a part of [this](https://github.com/ArmyCyberInstitute/cmgr/tree/master/examples#custom) would be helpful, especially the `PUBLISH` since the error generated really does not hint towards what could be the issue. Sample error log is as follows:

```
$ sudo ./cmgr update cert-pinning-1/
cmgr: [ERROR:  cannot use '{{port}}' in challenge type 'custom' which exposes 0 ports]
cmgr: [ERROR:  cannot use '{{server}}' in challenge type 'custom' which exposes 0 ports]
cmgr: [ERROR:  unrecognized template string of '{{port}}': /mnt/hgfs/picoCTF/dev/androidChallengeDevelopment/cert-pinning-1]
cmgr: [ERROR:  unrecognized template string of '{{server}}': /mnt/hgfs/picoCTF/dev/androidChallengeDevelopment/cert-pinning-1]
Errors:
    1) unrecognized template string of '{{server}}': /mnt/hgfs/picoCTF/dev/androidChallengeDevelopment/cert-pinning-1
```

Looking at the error log, I would presume that the error is with the `EXPOSE` rather than `PUBLISH`. 

For the last one, I downloaded the latest release available [here](https://github.com/ArmyCyberInstitute/cmgr/releases/latest/download/cmgr.tar.gz) and I can confirm that the issue persists. 

Apart from this, is there a way to remove a challenge that's been added into the `cmgr`? I don't know if I am missing out on something in the documentation, but I can't seem to figure that

**Comment by jrolli** [comment]
Thanks for the follow-up and I'll work on updating the documentation.

Regarding `CMGR_INTERFACE`, I think this is an issue with environment variables and `sudo`.  Specifically, `sudo` does not carry over new environment variables by default which is a more general problem.  Using `sudo -E ./cmgr playtest [...]` (passes all environment variables to sudo) or `sudo CMGR_INTERFACE=[ip addr] ./cmgr playtest [...]` (explicitly sets the environment variable in the subprocess) should fix the issue.

Finally, `cmgr update` will automatically remove challenges when three conditions are met:
1) It is called either without a path (implies `CMGR_DIR`) or with an explicity absolute or relative path that matches `CMGR_DIR`.
2) No challenge in `CMGR_DIR` matches the ID of the challenge.
3) There are no parsing/loading errors for any challenges (this prevents a mistake in updating the problem description from causing challenges to completely disappear).

In other words, it should be as simple as deleting the challenge's description (or the whole directory) and running `cmgr update`.

**Comment by jrolli** [comment]
@boomcan90, I've updated the documentation on master.  Is [this](https://github.com/ArmyCyberInstitute/cmgr/tree/master/examples#custom) documentation better and address your concerns?

**Comment by boomcan90** [comment]
Thanks for the info above. 

Could you also clarify what is the purpose of [this line](https://github.com/ArmyCyberInstitute/cmgr/blob/6110367dd891ff613153a44b8d2f69c1a7a7d182/examples/multi/Dockerfile#L1)? I could not find any references of `LAUNCH` within the dockerfile so I am assuming that it comes from `cmgr`.

**Comment by boomcan90** [comment]
Thanks for the above. That provides a lot more information and does address all my concerns!
